[Dynamo] Add UserError type

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #97705

To get started the dynamo error message improvement effort, we discussed about adding new user error type which covers cases where the user used something that TorchDynamo doesn't support and there is clear actions they can take.

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

tugsbayasgalan committed:
[Dynamo] Add UserError type

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[Dynamo] Add UserError type"


To get started the dynamo error message improvement effort, we discussed about adding new user error type which covers cases where the user used something that TorchDynamo doesn't support and there is clear actions they can take.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

anijain2305 commented on torch/_dynamo/exc.py:
```diff
@@ -73,6 +74,26 @@ def add_to_stats(self, category="unimplemented"):
+# supported. But seperate exception for now as to
+# not pollute the stat collector.
+class UserError(TorchDynamoException):
+    def __init__(self, error_type: UserErrorType, msg):
```
Please add comment here explaining UserError (same one that you have in the description).

tugsbayasgalan replied:
done

tugsbayasgalan committed:
Update on "[Dynamo] Add UserError type"


To get started the dynamo error message improvement effort, we discussed about adding new user error type which covers cases where the user used something that TorchDynamo doesn't support and there is clear actions they can take.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[Dynamo] Add UserError type"


To get started the dynamo error message improvement effort, we discussed about adding new user error type which covers cases where the user used something that TorchDynamo doesn't support and there is clear actions they can take.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[Dynamo] Add UserError type"


To get started the dynamo error message improvement effort, we discussed about adding new user error type which covers cases where the user used something that TorchDynamo doesn't support and there is clear actions they can take.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[Dynamo] Add UserError type"


To get started the dynamo error message improvement effort, we discussed about adding new user error type which covers cases where the user used something that TorchDynamo doesn't support and there is clear actions they can take.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot merge

tugsbayasgalan committed:
Update on "[Dynamo] Add UserError type"


To get started the dynamo error message improvement effort, we discussed about adding new user error type which covers cases where the user used something that TorchDynamo doesn't support and there is clear actions they can take.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot merge

voznesenskym commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -329,7 +329,12 @@ def inner(self: "InstructionTranslatorBase", inst: Instruction):
+            # TODO link the torch.cond doc later
+            raise exc.UserError(
+                exc.UserErrorType.DYNAMIC_CONTROL_FLOW,
+                "Dynamic control flow is not supported at the moment. Please use "
```
It's not dynamic control flow, it's data dependent control flow.
